src/block_manager/block_manager.rs

```diff
@@ -1,23 +1,60 @@
-use std::sync::Arc;
-use super::rlp_en_de::{encode_block_headers, encode_block_bodies};
-use crate::{client_adapter::Blockchain, scheduler::peer_organizer::{ErrorAct, Task}};
-use crate::scheduler::peer_organizer::PeerId;
-use crate::scheduler::protocol::{ProtocolId, MessageId, EthMessageId};
 // Copyright 2020 Gnosis Ltd.
 // SPDX-License-Identifier: Apache-2.0
 
-
-
-
+use std::sync::{Arc, Mutex};
+use super::rlp_en_de::{
+    decode_block_headers,
+    encode_block_headers,
+    encode_block_bodies,
+    encode_get_block_headers
+};
+use crate::{
+    client_adapter::Blockchain,
+    common_types::{BlockId, GetBlockHeaders},
+    scheduler::PeerOrganizer,
+    scheduler::peer_organizer::{ErrorAct, PeerId, Task},
+    scheduler::protocol::{ProtocolId, MessageId, EthMessageId}
+};
 
 pub struct BlockManager {
-    // chain: Arc<dyn Blockchain>,
+    chain: Arc<Mutex<dyn Blockchain + Send + Sync>>,
+    peer_organizer: Arc<Mutex<PeerOrganizer>>
 }
 
 
 
 //ALL APIs
 impl BlockManager {
+    pub fn new(
+        chain: Arc<Mutex<dyn Blockchain + Send + Sync>>,
+        peer_organizer: Arc<Mutex<PeerOrganizer>>
+    ) -> Arc<Mutex<Self>> {
+        Arc::new(Mutex::new(BlockManager { chain, peer_organizer }))
+    }
+
+    fn request_block_headers(&self, peer: &PeerId) -> Task {
+        let request = GetBlockHeaders::new(BlockId::Number(10_000_000), 100, 0, false);
+        let data = encode_get_block_headers(&request);
+        Task::GetBlocks(*peer, data)
+    }
+
+    pub fn sync(&self) {
```

Comment:
Hm this is probably why you needed peer_organizer. My thoughts on this are, you can create a Task without peer_id and send that to peer_organizer to be scheduled and inside of `schedule` function, you choose a random peer that will receive this task.